core/services/ocr2/plugins/ocr2keeper/evm21/core/payload.go

```diff
@@ -0,0 +1,86 @@
+package core
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
+)
+
+// UpkeepTriggerID returns the identifier using the given upkeepID and trigger.
+// It follows the same logic as the contract, but performs it locally.
+func UpkeepTriggerID(id *big.Int, trigger []byte) string {
+	idBytes := id.Bytes()
+
+	combined := append(idBytes, trigger...)
+
+	triggerIDBytes := crypto.Keccak256(combined)
+	triggerID := hex.EncodeToString(triggerIDBytes)
+
+	return triggerID
+}
+
+// UpkeepWorkID returns the identifier using the given upkeepID and trigger extension(tx hash and log index).
+func UpkeepWorkID(id *big.Int, trigger ocr2keepers.Trigger) (string, error) {
+	extensionBytes, err := json.Marshal(trigger.Extension)
```

Comment:
are we sure that the trigger is always a raw JSON?

Reply:
I'm not sure tbh, let's discuss offline. I would rather make extension a concrete type than have to worry about encoding/decoding

Reply:
discussed offline, will fix in followup PR to have a single entry point for trigger extension